Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+            }
+            dispatch_async(dispatch_get_main_queue()) {
+                self.collectionView.reloadData()
+            }
+        }
+    }
+
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        backgroundColor = ASHorizontalScrollCellUX.BackgroundColor
+        contentView.addSubview(collectionView)
+        contentView.addSubview(pageControl)
+
+        collectionView.snp_makeConstraints { make in
+            make.edges.equalTo(contentView).priorityLow()
+        }
+
+        pageControl.snp_makeConstraints { make in
+            make.size.equalTo(ASHorizontalScrollCellUX.PageControlSize)
+            make.top.equalTo(collectionView.snp_bottom).offset(ASHorizontalScrollCellUX.PageControlOffset)
+            make.centerX.equalTo(self.snp_centerX)
+        }
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        let layout = collectionView.collectionViewLayout as! HorizontalFlowLayout
+
+        pageControl.pageCount = layout.numberOfPages
+        pageControl.hidden = pageControl.pageCount <= 1
+    }
+
+    func currentPageChanged(currentPage: CGFloat) {
+        pageControl.progress = currentPage
+    }
+
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+/*
+    A custom layout used to show a horizontal scrolling list with paging. Similar to iOS springboard.
+    A modified version of http://stackoverflow.com/a/34167915
+ */
+
+class HorizontalFlowLayout: UICollectionViewLayout {
+    var itemSize = CGSize.zero
+    private var cellCount = 0
+    private var boundsSize = CGSize.zero
+    private var insets = UIEdgeInsetsZero
+    private let minimumInsets: CGFloat = 20
+    var numberOfPages = 0
+
+    override func prepareLayout() {
+        cellCount = self.collectionView!.numberOfItemsInSection(0)
+        boundsSize = self.collectionView!.bounds.size
+    }
+
+    override func collectionViewContentSize() -> CGSize {
+        return collectionViewSizeForRect(boundsSize)
+    }
+
+    func maxVerticalItemsCount() -> Int {
+        let verticalItemsCount =  Int(floor(boundsSize.height / (itemSize.height + insets.top)))
+        if let delegate = self.collectionView?.delegate as? ASHorizontalScrollDelegate {
+            return delegate.numberOfVerticalItems(verticalItemsCount)
+        } else {
+            return verticalItemsCount
+        }
+    }
+
+    func maxHorizontalItemsCount() -> Int {
+        let horizontalItemsCount =  Int(floor(boundsSize.width / (itemSize.width + insets.left)))
+        if let delegate = self.collectionView?.delegate as? ASHorizontalScrollDelegate {
+            return delegate.numberOfHorizontalItems(horizontalItemsCount)
+        } else {
+            return horizontalItemsCount
+        }
+    }
+    func collectionViewSizeForRect(contentSize: CGSize) -> CGSize {
```

Comment:
Nit: Since this method is only invoked from `collectionViewContentSize`, can we just remove this method and put the logic in that?